### ISSUE TITLE
Pushdown semijoin below agg if keys are subset

### ISSUE THREAD

**Issue by kaikalur** [prestodb/presto#23522]
select .. from 
(select agg.. from t group by k,l)
where k in (select distinct k from t2)

will produce a plan that does the semijoin after the full agg. But semijoin is cardinality reducing (or perserving at worst) so we should push it down. So we should do the equivalent of:

select .. 
(select agg from t where k in (select distinct k from t2) group by k,l)

Bonus (from prod):

select .. from 
(select agg.. from t group by k,l union all select agg.. from t1 group by k,l)

to:
select .. from 
(select agg.. from t  where k in (select distinct k from t2) group by k,l 
union all select agg.. from t1  where k in (select distinct k from t2) group by k,l)

basically distribute semijoin into union all and then apply pushdown to below agg (partial)
